File: services/email_service.py
"""Email service with HTML templates and custom SMTP configuration."""

import logging
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

from django.template import Template, Context

logger = logging.getLogger(__name__)


class EmailService:
    """Service for sending emails using DB-configured SMTP and templates."""

    def _get_smtp_config(self):
        """Get the active SMTP configuration from the database."""
        from apps.accounts.models import SMTPConfiguration
        config = SMTPConfiguration.objects.filter(is_active=True).first()
        if not config:
            raise ValueError("Aucune configuration SMTP active trouvée")
        return config

    def _get_template(self, template_type):
        """Get an email template by type."""
        from apps.accounts.models import EmailTemplate
        template = EmailTemplate.objects.filter(
            template_type=template_type, is_active=True
        ).first()
        if not template:
            raise ValueError(f"Template email '{template_type}' introuvable ou inactif")
        return template

    def _render_template(self, template_str, context_dict):
        """Render a Django template string with the given context."""
        template = Template(template_str)
        context = Context(context_dict)
        return template.render(context)

    def _send_email(self, smtp_config, to_email, subject, html_body):
        """Send an email using the given SMTP configuration."""
        msg = MIMEMultipart('alternative')
        msg['Subject'] = subject
        msg['From'] = f"{smtp_config.from_name} <{smtp_config.from_email}>"
        msg['To'] = to_email

        html_part = MIMEText(html_body, 'html', 'utf-8')
        msg.attach(html_part)

        if smtp_config.use_ssl:
            server = smtplib.SMTP_SSL(smtp_config.host, smtp_config.port, timeout=10)
        else:
            server = smtplib.SMTP(smtp_config.host, smtp_config.port, timeout=10)

        try:
            if smtp_config.use_tls and not smtp_config.use_ssl:
                server.starttls()
            if smtp_config.username:
                server.login(smtp_config.username, smtp_config.password)
            server.sendmail(smtp_config.from_email, [to_email], msg.as_string())
            logger.info("Email sent to %s: %s", to_email, subject)
        finally:
            server.quit()

    def send_template_email(self, template_type, to_email, context_dict=None):
        """Send an email using a stored template."""
        smtp_config = self._get_smtp_config()
        template = self._get_template(template_type)

        context_dict = context_dict or {}
        from django.conf import settings
        context_dict.setdefault('site_name', settings.SITE_NAME)
        context_dict.setdefault('site_url', settings.SITE_URL)

        subject = self._render_template(template.subject, context_dict)
        html_body = self._render_template(template.html_body, context_dict)

        self._send_email(smtp_config, to_email, subject, html_body)

    def send_welcome(self, user):
        """Send welcome email to a new user."""
        self.send_template_email('welcome', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
        })

    def send_payment_reminder(self, user, subscription, days_overdue):
        """Send payment reminder email."""
        self.send_template_email('payment_reminder', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
            'subscription': subscription,
            'days_overdue': days_overdue,
        })

    def send_password_reset(self, user, reset_url):
        """Send password reset email."""
        self.send_template_email('password_reset', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
            'reset_url': reset_url,
        })

    def send_account_suspended(self, user):
        """Send account suspended notification."""
        self.send_template_email('account_suspended', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
        })

    def send_account_deleted(self, user):
        """Send account deleted notification."""
        self.send_template_email('account_deleted', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
        })

    def send_gift_received(self, recipient_email, gifted_by, plan):
        """Send gift received notification."""
        self.send_template_email('gift_received', recipient_email, {
            'gifted_by': gifted_by,
            'plan': plan,
        })

    def send_payment_success(self, user, payment):
        """Send payment success notification."""
        self.send_template_email('payment_success', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
            'payment': payment,
        })

    def send_refund_processed(self, user, refund):
        """Send refund processed notification."""
        self.send_template_email('refund_processed', user.email, {
            'user': user,
            'username': user.username,
            'first_name': user.first_name,
            'refund': refund,
        })

    def test_smtp_config(self, smtp_config, test_email):
        """Test SMTP configuration by sending a test email."""
        try:
            self._send_email(
                smtp_config,
                test_email,
                'Test de configuration SMTP — MonFlow',
                '<h1>Test SMTP</h1><p>La configuration SMTP fonctionne correctement.</p>',
            )
            return True, "Email de test envoyé avec succès"
        except Exception as e:
            return False, f"Erreur: {e}"


email_service = EmailService()

File: app/Services/EmailService.php
<?php

namespace App\Services;

use App\Models\{SmtpConfiguration, EmailTemplate, EmailLog, User};
use App\Jobs\SendEmailJob;
use Illuminate\Support\Facades\{Log, URL};

class EmailService
{
    private function getSmtp(): SmtpConfiguration
    {
        return SmtpConfiguration::where('is_active', true)->firstOrFail();
    }

    private function render(string $template, array $ctx): string
    {
        foreach ($ctx as $k => $v) {
            $template = str_replace("{{ {$k} }}", (string) $v, $template);
            $template = str_replace("{{{$k}}}", (string) $v, $template);
        }
        return $template;
    }

    private function send(SmtpConfiguration $smtp, string $to, string $subject, string $html, ?string $type = null, array $attachments = []): void
    {
        $transport = new \Symfony\Component\Mailer\Transport\Smtp\EsmtpTransport(
            $smtp->host, $smtp->port, $smtp->use_tls
        );
        if ($smtp->username) {
            $transport->setUsername($smtp->username);
            $transport->setPassword($smtp->password);
        }
        $mailer = new \Symfony\Component\Mailer\Mailer($transport);
        $email = (new \Symfony\Component\Mime\Email())
            ->from(new \Symfony\Component\Mime\Address($smtp->from_email, $smtp->from_name))
            ->to($to)
            ->subject($subject)
            ->html($html);

        foreach ($attachments as $attachment) {
            $email->attachFromPath($attachment['path'], $attachment['name'] ?? null, $attachment['mime'] ?? null);
        }

        try {
            $mailer->send($email);
            Log::info("Email sent to {$to}: {$subject}");
            EmailLog::create(['to' => $to, 'subject' => $subject, 'template_type' => $type, 'html_body' => $html, 'status' => 'sent']);
        } catch (\Throwable $e) {
            EmailLog::create(['to' => $to, 'subject' => $subject, 'template_type' => $type, 'html_body' => $html, 'status' => 'failed', 'error' => $e->getMessage()]);
            throw $e;
        }
    }

    public function sendTemplate(string $type, string $toEmail, array $ctx = []): void
    {
        SendEmailJob::dispatch($type, $toEmail, $ctx);
    }

    public function sendTemplateNow(string $type, string $toEmail, array $ctx = []): void
    {
        try {
            $smtp = $this->getSmtp();
            $tpl = EmailTemplate::where('template_type', $type)->where('is_active', true)->firstOrFail();
            $ctx['site_name'] = config('app.name');
            $ctx['site_url'] = config('app.url');
            $subject = $this->render($tpl->subject, $ctx);
            $body = $this->render($tpl->html_body, $ctx);
            $this->send($smtp, $toEmail, $subject, $body, $type);
        } catch (\Exception $e) {
            Log::error("Failed to send email [{$type}] to {$toEmail}: {$e->getMessage()}");
        }
    }

    /**
     * Envoi synchrone avec pièce jointe (rapport URSSAF) : contrairement à
     * sendTemplateNow(), les erreurs remontent pour que l'appelant puisse
     * marquer l'envoi en échec plutôt que le considérer silencieusement fait.
     */
    public function sendUrssafReport(string $toEmail, string $periodLabel, float $total, int $count, string $pdfPath): void
    {
        $smtp = $this->getSmtp();
        $tpl = EmailTemplate::where('template_type', 'urssaf_report')->where('is_active', true)->firstOrFail();
        $ctx = [
            'period' => $periodLabel,
            'total' => number_format($total, 2, ',', ' '),
            'count' => $count,
            'site_name' => config('app.name'),
            'site_url' => config('app.url'),
        ];
        $subject = $this->render($tpl->subject, $ctx);
        $body = $this->render($tpl->html_body, $ctx);
        $this->send($smtp, $toEmail, $subject, $body, 'urssaf_report', [
            ['path' => $pdfPath, 'name' => "declaration-urssaf-{$periodLabel}.pdf", 'mime' => 'application/pdf'],
        ]);
    }

    public function sendVerification(User $u, string $url): void { $this->sendTemplate('email_verification', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'verify_url' => $url]); }
    public function sendWelcome(User $u): void { $this->sendTemplate('welcome', $u->email, ['username' => $u->username, 'first_name' => $u->first_name]); }
    public function sendPaymentReminder(User $u, int $days): void { $this->sendTemplate('payment_reminder', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'days_overdue' => $days]); }
    public function sendPasswordReset(User $u, string $url): void { $this->sendTemplate('password_reset', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'reset_url' => $url]); }
    public function sendSuspended(User $u): void { $this->sendTemplate('account_suspended', $u->email, ['username' => $u->username, 'first_name' => $u->first_name]); }
    public function sendDeleted(User $u): void
    {
        // Lien signé (30 jours) permettant à l'utilisateur de libérer lui-même
        // son ancienne adresse email pour se réinscrire, sans avoir à contacter
        // le support ni à s'authentifier (le compte est supprimé).
        $resubscribeUrl = URL::temporarySignedRoute('resubscribe', now()->addDays(30), ['id' => $u->id]);
        $this->sendTemplate('account_deleted', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'resubscribe_url' => $resubscribeUrl]);
    }
    public function sendDeletedRecoverable(User $u, float $fee): void { $this->sendTemplate('account_deleted_recoverable', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'fee' => number_format($fee, 2, ',', ' ')]); }
    public function sendDeletionWarning(User $u, int $daysLeft, \DateTimeInterface $deletionDate): void { $this->sendTemplate('deletion_warning', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'days_left' => $daysLeft, 'deletion_date' => $deletionDate->format('d/m/Y')]); }
    public function sendGiftReceived(string $email, string $plan): void { $this->sendTemplate('gift_received', $email, ['plan' => $plan]); }
    public function sendRefund(User $u): void { $this->sendTemplate('refund_processed', $u->email, ['username' => $u->username, 'first_name' => $u->first_name]); }
    public function sendRenewalReminder(User $u, \App\Models\Plan $plan, float $price, bool $promoEnding = false): void { $this->sendTemplate('renewal_reminder', $u->email, ['username' => $u->username, 'first_name' => $u->first_name, 'plan_name' => $plan->name, 'price' => $price, 'promo_ending' => $promoEnding]); }

    public function sendNewsletterNow(User $u, string $subject, string $htmlBody): void
    {
        $smtp = $this->getSmtp();
        $ctx = ['username' => $u->username, 'first_name' => $u->first_name, 'site_name' => config('app.name'), 'site_url' => config('app.url'), 'sujet' => $subject];
        $renderedSubject = $this->render($subject, $ctx);
        $renderedBody = $this->render($htmlBody, $ctx);
        $unsubLink = config('app.url') . '/portal/profile';
        $renderedBody = str_replace('</body>', "<div style=\"text-align:center;padding:16px;font-size:11px;color:#a1a1aa\"><a href=\"{$unsubLink}\" style=\"color:#6366f1\">Se désabonner de la newsletter</a></div></body>", $renderedBody);
        $this->send($smtp, $u->email, $renderedSubject, $renderedBody, 'newsletter');
    }

    public function testSmtp(SmtpConfiguration $smtp, string $testEmail): array
    {
        try {
            $this->send($smtp, $testEmail, 'Test SMTP — ' . config('app.name'), '<h1>Test OK</h1><p>La configuration SMTP fonctionne.</p>', 'test_smtp');
            return ['success' => true, 'message' => 'Email envoyé avec succès'];
        } catch (\Exception $e) {
            return ['success' => false, 'message' => $e->getMessage()];
        }
    }
}
